README.md

```diff
@@ -1,2 +1,159 @@
+those OS tasks are performed for each of them.
+
+The execution time of this shell script is around **180-230ms on average**.
+
+However, if I, functionality-wise. implement, the exact same script but in C in which I compile it all in
+a single binary, I can get the execution time to be around **5-13ms on average**. 
+
+Creating this `'monolith-binary'` is a well-known technique that embedded developers tend to use. It saves both storage
+and memory space for low-resource devices. (Busybox and Toybox are well-known libraries that implement such approach).
+
+While I'm not aiming this idea for low-resource devices, it also happens to be a way to reduce execution time of a
+program (by eliminating most of the housekeeping OS tasks I mentioned earlier).
+
+
+# proposed solution
+This idea aims for reducing the execution times disparity between bash script and monolith-binaries for high-resource
+devices such as webservers and HPC clusters. It has 2 big components to make it work as shown below:
+
+## kernel space - new syscalls
+
+Currently, syscall `execve(2)` initiates a binary execution in the kernel. 
+
+From the [docs](https://man7.org/linux/man-pages/man2/execve.2.html):
+> execve() executes the program referred to by pathname.  This <br/>
+> causes the program that is currently being run by the calling <br/>
+> process to be replaced with a new program, with newly initialized <br/>
+> stack, heap, and (initialized and uninitialized) data segments.
+
+This triggers a bunch of userspace <-> kernel interactions that eventually
+run the program passed in as a parameter. The problem with that is:
+
+1. It's an I/O blocking operation
+2. Once triggered the program will run, there is no way back. That's equivalent
+ of droping down a pebble at the top of a cliff
+
+What I propose is to create 2 new syscalls:
+
+- `prewarm`: which will do pretty much everything that `execve` does however, it will
+   set a new task state to [`TASK_NEW`](https://elixir.bootlin.com/linux/v6.1.1/source/include/linux/sched.h#L100)
+   on the OS scheduler so the process is never scheduled for execution until `runpid`
+   syscall is triggered. (This would leverage the deferred-executions in the kernel to make it
+   as async as possible).
+
+- `runpid`: which will change existing task state to [`TASK_RUNNING`](https://elixir.bootlin.com/linux/v6.1.1/source/include/linux/sched.h#L85)
+   which essentially tells the OS scheduler that it can run at scheduler's earliest
+   convinience.
+
+PS: names of syscalls are most definitely open for suggestions, I struggle with giving 
+things code names. I'm horrible at this :-)
+
+## user space - bash interpreter
+
+My idea is to add a `shopt` option that once enabled, it would instruct bash interpreter
+that upload script loading, it would:
+
+1. extract all commands to be sent to the OS from its parser grammar. Bash already parses it's
```

Comment:
This problem sounds similar to what nushell solved differently (I think?) Instead of tokenizing the script and parsing the commands out, they instead provide a shell with the "same" commands that are more efficiently run (by some means they control).

Reply:
My 2 cents is that nushell tackled a different issue (no input/output data standardisation). The fact that they are constrained to user-space only, limits the performance gain they can yield. 

For instance, they get some performance gain due to the shell pipeline re-implementation and internal job control but the way and order in which processes are loaded, scheduled and executed is essentially the same.

TBH, I think that if this idea succeeds then nushell would get an additional performance gain out-of-the-box :)